Enable parallel builds by default

diizzyy opened the pull request:
Enable parallel builds by default and reset all packages including those in tools directory

diizzyy committed:
treewide: Remove PKG_BUILD_PARALLEL for all packages

Enable parallel builds for all packages by default

Signed-off-by: Daniel Engberg <daniel.engberg.lists@pyret.net>

diizzyy committed:
toolchain: Enable parallel builds by default

About any software is developed or at least tested with paralell builds and
the majority of all package systems have this enabled by default to speed up
compilation. This change also applies to host builds.

Signed-off-by: Daniel Engberg <daniel.engberg.lists@pyret.net>

diizzyy committed:
toolchain: Remove HOST_BUILD_PARALLEL for all packages in tools directory

No need to specify HOST_BUILD_PARALLEL as it's enabled by default

Signed-off-by: Daniel Engberg <daniel.engberg.lists@pyret.net>

diizzyy commented:
Compile/Run tested: ar71xx, TP-Link TL-WR810N v2, OpenWrt HEAD

diizzyy commented:
I've tested various packages and they all seem fine but I'm sure we'll run into some edge case at some point.

champtar commented:
@diizzyy I think you need to run at least 1 full build with ALL packages if you want to have a chance for this change to be considered

karlp commented on package/libs/gettext-full/Makefile:
```diff
@@ -23,7 +23,6 @@ PKG_CPE_ID:=cpe:/a:gnu:gettext
 
 PKG_FIXUP:=autoreconf
 PKG_INSTALL:=1
-PKG_BUILD_PARALLEL:=0
```
So this is a package that is explicitly setting it to 0.  I'm all for removing the =1 where that matches the new default, but I think changing the setting of a package that has already got to the explicit trouble of disabling parallel builds should be left alone _for this commit_ at least.  By all means separately as a follow up see if it still has problems, but it seems hazardous to blend it into this large commit

diizzyy replied:
Can't reproduce any build issues

karlp commented on package/libs/libtool/Makefile:
```diff
@@ -19,8 +19,6 @@ PKG_LICENSE:=GPL-2.0+
 PKG_CPE_ID:=cpe:/a:gnu:libtool
 
-PKG_BUILD_PARALLEL:=0
-
```
here too.

diizzyy replied:
Can't reproduce any build issues

karlp replied:
I don't really care whether you can or not, I think it's just simply out of scope of this PR, and your single point test environment is simply not sufficient.  (Same comment applies to the rest of them)

diizzyy replied:
If you don't care why even bother commenting?
I'll do builds locally (it'll take some time) and then I'd suggest throwing it on a _reliable_ buildbot as at least one seems dodgy looking at the faillogs.

karlp replied:
I just didn't word it well enough, my apologies.  I just don't think a single build that compiled ok is evidence that these packages are safe.  I _very much_ want to see this patch land to turn on parallel builds by default for packages with no preference.  I very strongly believe that changing the setting for packages that DO have a preference is out of scope for _this_ commit, and should be done in a follow up.

karlp commented on package/libs/libusb/Makefile:
```diff
@@ -18,7 +18,6 @@ PKG_SOURCE_URL:=\
 PKG_HASH:=75aeb9d59a4fdb800d329a545c2e6799f732362193b465ea198f2aa275518157
 
 PKG_INSTALL:=1
-PKG_BUILD_PARALLEL:=0
```
and this one. I think all of these should be a follow up PR later

diizzyy replied:
Can't reproduce any build issues

karlp commented on package/libs/nettle/Makefile:
```diff
@@ -18,7 +18,6 @@ PKG_HASH:=ae7a42df026550b85daca8389b6a60ba6313b0567f374392e54918588a411e94
 
 PKG_LICENSE:=GPL-2.0+
 PKG_LICENSE_FILES:=COPYING
-PKG_BUILD_PARALLEL:=0
```
one more

diizzyy replied:
Can't reproduce any build issues

karlp commented on package/libs/openssl/Makefile:
```diff
@@ -14,9 +14,6 @@ PKG_VERSION:=$(PKG_BASE)$(PKG_BUGFIX)
 
-PKG_BUILD_PARALLEL:=0
-
-
```
and this one

diizzyy replied:
Can't reproduce any build issues
It does however revert to -j1 in some parts of the build, I don't think it's worth spending much time on as we'll move to 1.1.1 soon.

karlp commented on package/system/uci/Makefile:
```diff
@@ -22,8 +22,6 @@ PKG_LICENSE_FILES:=
 
 PKG_MAINTAINER:=Felix Fietkau <nbd@nbd.name>
 
-PKG_BUILD_PARALLEL:=0
```
here too

diizzyy replied:
Can't reproduce any build issues

karlp commented:
I like this a lot apart from removing the flag on the explicit packages as noted.

diizzyy commented:
I tried to make a full build but documentation is sparse about logging and it does fail at the end which however doesn't seem to be related to this change as ar71xx seems to failed on the buildbots for a couple of days.

```make -j4 V=s IGNORE_ERRORS=m BUILD_LOG=1 2>&1 | tee build.log```
http://projects.pyret.net/dump/lede/build-ar71xx.zip

@karlp 
The ones you pointed out are tested with -j6

nbd168 commented:
I think testing a full build with -j4 is simply not enough to check for individual package parallel build issues. To be certain, you should test each package with `make -j8 package/<foo>/{clean,compile}`.

diizzyy committed:
toolchain: Remove HOST_BUILD_PARALLEL for all packages in toolchain directory

No need to specify HOST_BUILD_PARALLEL as it's enabled by default

Signed-off-by: Daniel Engberg <daniel.engberg.lists@pyret.net>

diizzyy committed:
conntrack-tools: Disable parallel build

Disable parallel build as compilation gives the following error:
read_config_lex.l:25:10: fatal error: read_config_yy.h: No such file or directory

Signed-off-by: Daniel Engberg <daniel.engberg.lists@pyret.net>

diizzyy commented:
I've tested **-j8** on all packages except a very few platform specific ones (I used mips 24kc as target) and they all work fine for me (tm) except contrack-tools.

This includes:
```
./package
./toolchain
./tools
```

@nbd168 
You can perhaps do a buildbot run?

LGA1150 commented:
Tried with `-j16`. Encountered random compile errors when enabled by default. Most common errors are `Not such file or directory` or even `Segmentation fault`

diizzyy commented:
Target? Logs?

diizzyy commented:
I tried -j8 on packages and -j16 on full build run

ldir-EDB0 commented:
IIRC some packages have parallel build disabled for binary reproducibility reasons.  They may well compile fine, but the resultant binary isn't bit identical.

bobafetthotmail commented:
> IIRC some packages have parallel build disabled for binary reproducibility reasons. They may well compile fine, but the resultant binary isn't bit identical.

In that case it's probably better to have this as an option you can enable on your local builds.

diizzyy commented:
@ldir-EDB0 
Can you provide anything that backs up that claim?

karlp commented:
> @ldir-EDB0
> Can you provide anything that backs up that claim?

Why not just take as it _yet more_ evidence of a desire to get parallel builds turned on by default, but to leave changing packages that already took an explicit decision to more focused issues, evaluating them one by one.

ldir-EDB0 commented:
Sadly no, which suggests my memory is failing me.  git log -p  -S PKG_BUILD_PARALLEL:=0 shows me a number of packages that have been switched to non-parallel for build/linking issues but not specifically for reproducibility - I've pinged lynxis as he may remember better the parallel/reproducible issues as he's been fighting that battle.

I'm fine with enabling parallel by default.  I'm not so fine with the same PR removing BUILD_PARALLEL:=0

diizzyy commented:
@ldir-EDB0 
Yeah, the lack of documentation or verbose commit messages doesn't really help :/
I do get the impression that most (all?= are added due to race conditions in older versions.

diizzyy commented:
This is going nowhere (unsurprisingly) so might as well close it
